File: node/src/runtime/reactive-session.ts
/**
 * L2 (ReactiveSession) — the user-facing primitive for "N peer agents over a shared event stream"
 * (spec §6). It composes the lower layers so teams don't hand-roll the pattern:
 *   - L1 `RunGroup`     — shared governance domain (cumulative budget + lineage) across the personas.
 *   - L0 `SignalGateway`— recipient-routed signals (targeted `interrupt` / `broadcast`).
 *   - `EventStream`     — the shared blackboard (pluggable; default in-memory).
 *   - `TurnPolicy`      — who reacts to each event (the one caller-customizable seam).
 *
 * Stateless-friendly: `emit` can run inside an HTTP handler; each persona's turn is a normal
 * `run({sessionId})` whose continuity comes from its `SessionLog`, and `resume()` rebuilds the peer
 * set from the persisted `RunGroup` membership — no hot in-process loop required.
 */
import type { RuntimeRunner } from "./runner.js"
import { collectText } from "./runner.js"
import type { RunGroup } from "./run-group.js"
import type { SignalSource, RuntimeSignal } from "../signals/types.js"
import { SignalGateway } from "../os/public.js"
import type { BlackboardEvent, EventStream, EventViewer } from "./event-stream.js"
import { InMemoryEventStream, isVisibleTo } from "./event-stream.js"
import type { TurnPolicy, PeerView } from "./turn-policy.js"
import { tool } from "../tools/index.js"
import type { RegisteredTool } from "../tools/index.js"

/**
 * How a persona executes one reactive turn. The default body is a single `runner.run(...)` agent turn;
 * override it to make a persona's turn a *different orchestration form* — e.g. drive a DAG via
 * `ctx.runner.runWorkflow(spec)` (DAG-in-Peer) or any composite. The runner is already wired to the
 * shared `RunGroup`, so whatever the body spawns stays under one governance domain. Must return the
 * persona's reaction text.
 */
export interface ReactorContext {
  personaId: string
  goal: string
  event: BlackboardEvent
  /** The persona's runner — wired to the shared RunGroup / signal gateway / blackboard. */
  runner: RuntimeRunner
}
export type ReactorTurn = (ctx: ReactorContext) => Promise<string>

/** Per-persona registration: its base reaction goal, role, channel subscriptions, and turn body. */
export interface ReactivePeerSpec {
  goal?: string
  role?: string
  channels?: string[]
  /**
   * Override this persona's turn body (the seam for composing other mechanisms into a peer). Defaults
   * to the session `reactWith`, then to a single `run()` agent turn. Use to make this peer's reaction
   * a workflow DAG, a nested ensemble, etc. — all under the shared `RunGroup`.
   */
  react?: ReactorTurn
}

/** What the caller appends to the blackboard via `emit`. */
export interface EmitEvent {
  payload: unknown
  source?: string
  channel?: string
  audience?: string[]
}

export interface ReactiveSessionOptions {
  /** Shared governance domain — all personas run under it (L1). */
  runGroup: RunGroup
  /** Who reacts to each event (L2). */
  turnPolicy: TurnPolicy
  /** Shared blackboard. Defaults to a process-local `InMemoryEventStream`. */
  eventStream?: EventStream
  /** Shared signal gateway for targeted interrupt / broadcast (L0). Defaults to a fresh one. */
  signalGateway?: SignalGateway
  /**
   * Build a runner for a persona, wiring in the shared governance + signal routing. The app owns the
   * provider / execution plane / tools; spread `shared` into the `RuntimeRunner` options and register
   * `readRecentTool(shared.eventStream, viewer)` so the persona can read the blackboard.
   */
  makeRunner: (
    personaId: string,
    shared: { runGroup: RunGroup; signalSource: SignalSource; eventStream: EventStream },
  ) => RuntimeRunner
  /** Goal for a persona's reactive turn. Defaults to a generic "react to the blackboard" prompt. */
  goalFor?: (personaId: string, event: BlackboardEvent) => string
  /**
   * Default turn body for peers that don't set their own `react`. Defaults to a single `run()` agent
   * turn. Override to make every peer's turn a different orchestration form (e.g. a workflow DAG).
   */
  reactWith?: ReactorTurn
}

/** A persona's reaction to an emitted event. */
export interface Reaction {
  personaId: string
  output: string
}

export class ReactiveSession {
  private readonly peerSpecs = new Map<string, ReactivePeerSpec>()
  private readonly runners = new Map<string, RuntimeRunner>()
  private readonly policyState: Record<string, unknown> = {}
  private readonly eventStream: EventStream
  private readonly gateway: SignalGateway

  constructor(private readonly opts: ReactiveSessionOptions) {
    this.eventStream = opts.eventStream ?? new InMemoryEventStream()
    this.gateway = opts.signalGateway ?? new SignalGateway()
  }

  /** Register a peer persona and record it in the group membership (lineage). */
  addPeer(personaId: string, spec: ReactivePeerSpec = {}): void {
    this.peerSpecs.set(personaId, spec)
    // W-N5: tagged "peer" so resume() can tell personas apart from vehicle sessions (workflow
    // envelopes / wf-node children / loop iterations) that share the same governance domain.
    void this.opts.runGroup.budgetStore.join(this.opts.runGroup.id, {
      sessionId: personaId,
      role: spec.role,
      kind: "peer",
    })
  }

  peers(): string[] {
    return [...this.peerSpecs.keys()]
  }

  blackboard(): EventStream {
    return this.eventStream
  }

  /**
   * Append an event to the blackboard, ask the `TurnPolicy` which (visible) peers react, and drive one
   * turn for each — returning their outputs. Each turn runs under the shared `RunGroup` governance.
   */
  async emit(event: EmitEvent): Promise<Reaction[]> {
    const bbEvent = await this.eventStream.append(event)

    const candidates: PeerView[] = [...this.peerSpecs.entries()]
      .map(([personaId, spec]) => ({ personaId, role: spec.role, channels: spec.channels }))
      // Only personas that can actually see the event are eligible to react.
      .filter(p => isVisibleTo(bbEvent, p as EventViewer))

    const chosen = await this.opts.turnPolicy(bbEvent, candidates, this.policyState)
    const eligible = new Set(candidates.map(p => p.personaId))

    const reactions: Reaction[] = []
    for (const personaId of chosen) {
      if (!eligible.has(personaId)) continue
      reactions.push({ personaId, output: await this.driveTurn(personaId, bbEvent) })
    }
    return reactions
  }

  /** Targeted preemption: deliver a critical signal to one persona's loop only (L0 recipient routing). */
  async interrupt(personaId: string, signal: Partial<RuntimeSignal> & { payload?: Record<string, unknown> }): Promise<void> {
    this.gateway.ingest({
      source: "gateway",
      signalType: "alert",
      urgency: "critical",
      payload: signal.payload ?? {},
      ...signal,
      recipient: personaId,
    })
  }

  /** Broadcast a signal to every persona (each sees it on its next turn). */
  async broadcast(signal: Partial<RuntimeSignal> & { payload?: Record<string, unknown> }): Promise<void> {
    this.gateway.ingest({
      source: "gateway",
      signalType: "event",
      urgency: "normal",
      payload: signal.payload ?? {},
      ...signal,
      recipient: undefined,
    })
  }

  private getRunner(personaId: string): RuntimeRunner {
    let runner = this.runners.get(personaId)
    if (!runner) {
      runner = this.opts.makeRunner(personaId, {
        runGroup: this.opts.runGroup,
        signalSource: this.gateway,
        eventStream: this.eventStream,
      })
      this.runners.set(personaId, runner)
    }
    return runner
  }

  private async driveTurn(personaId: string, event: BlackboardEvent): Promise<string> {
    const runner = this.getRunner(personaId)
    const goal =
      this.opts.goalFor?.(personaId, event) ??
      this.peerSpecs.get(personaId)?.goal ??
      "React to the latest events on the shared blackboard."
    // Turn-body seam (the DAG-in-Peer enabler): a persona's reaction can be any orchestration form,
    // not just a single agent turn. Per-peer `react` wins, then session `reactWith`, else the default
    // `run()`. Whatever the body drives (e.g. `runner.runWorkflow`) inherits the shared RunGroup.
    const react = this.peerSpecs.get(personaId)?.react ?? this.opts.reactWith
    if (react) return react({ personaId, goal, event, runner })
    // run() with the persona's stable sessionId replays its prior turns from the SessionLog, so this
    // is continuity-preserving whether or not the persona has acted before (stateless-handler safe).
    return collectText(runner.run({ sessionId: personaId, goal }))
  }

  /**
   * Rebuild a session from a persisted `RunGroup`: load its PEER members (lineage) as peers. The
   * blackboard continuity comes from the (persistent) `EventStream`. Turn-policy cursor state is not
   * restored. W-N5: vehicle members (workflow envelopes, `wf-node*` children, loop iterations) share
   * the governance domain but are NOT personas — resuming them as peers would resurrect phantoms.
   * A legacy membership with no kind tags falls back to resuming every member.
   */
  static async resume(
    opts: ReactiveSessionOptions & { peerSpecs?: Record<string, ReactivePeerSpec> },
  ): Promise<ReactiveSession> {
    const session = new ReactiveSession(opts)
    const members = await opts.runGroup.budgetStore.members(opts.runGroup.id)
    const anyTagged = members.some(m => m.kind !== undefined)
    for (const member of members) {
      if (anyTagged && member.kind !== "peer") continue
      session.peerSpecs.set(member.sessionId, opts.peerSpecs?.[member.sessionId] ?? { role: member.role })
    }
    return session
  }
}

/**
 * A `read_recent` tool a persona uses to read the shared blackboard, scoped to what it may see. Register
 * one per persona inside `makeRunner`. `viewer` is the reading persona (id + subscribed channels).
 */
export function readRecentTool(eventStream: EventStream, viewer: EventViewer): RegisteredTool {
  return tool(
    "read_recent",
    "Read recent events from the shared blackboard visible to you (optionally a single channel).",
    {
      type: "object",
      properties: {
        since_seq: { type: "number", description: "Only events after this seq (default: from the start)." },
        channel: { type: "string", description: "Restrict to one channel you subscribe to." },
      },
    },
    async (args) => {
      const sinceSeq = typeof args.since_seq === "number" ? args.since_seq : -1
      const channel = typeof args.channel === "string" ? args.channel : undefined
      const events = await eventStream.readSince(sinceSeq, viewer)
      const filtered = channel ? events.filter(e => e.channel === channel) : events
      return JSON.stringify(filtered.map(e => ({ seq: e.seq, source: e.source, channel: e.channel, payload: e.payload })))
    },
  )
}

File: python/deepstrike/runtime/reactive_session.py
"""L2 (ReactiveSession) — the user-facing primitive for "N peer agents over a shared event stream".

Composes the lower layers so teams don't hand-roll the pattern:
  - L1 ``RunGroup``      — shared governance domain (cumulative budget + lineage) across the personas.
  - L0 ``SignalGateway`` — recipient-routed signals (targeted ``interrupt`` / ``broadcast``).
  - ``EventStream``      — the shared blackboard (pluggable; default in-memory).
  - ``TurnPolicy``       — who reacts to each event (the one caller-customizable seam).

Stateless-friendly: ``emit`` can run inside an HTTP handler; each persona's turn is a normal
``run(session_id=...)`` whose continuity comes from its ``SessionLog``, and ``resume`` rebuilds the
peer set from the persisted ``RunGroup`` membership — no hot in-process loop required.
"""

from __future__ import annotations

import json
from dataclasses import dataclass, field
from typing import TYPE_CHECKING, Any, Awaitable, Callable

from deepstrike.runtime.event_stream import BlackboardEvent, EventStream, EventViewer, InMemoryEventStream, is_visible_to
from deepstrike.runtime.run_group import GroupMember, RunGroup
from deepstrike.runtime.runner import collect_text
from deepstrike.runtime.turn_policy import PeerView, TurnPolicy
from deepstrike.signals.gateway import SignalGateway
from deepstrike.signals.types import RuntimeSignal
from deepstrike.tools import tool

if TYPE_CHECKING:
    from deepstrike.runtime.runner import RuntimeRunner


@dataclass
class ReactorContext:
    """Context for one reactive turn. ``runner`` is wired to the shared RunGroup / signal gateway /
    blackboard, so whatever the turn body drives (e.g. ``runner.run_workflow``) stays under one
    governance domain."""
    persona_id: str
    goal: str
    event: BlackboardEvent
    runner: "RuntimeRunner"


# A turn body: given the context, return the persona's reaction text. Override to make a peer's turn a
# different orchestration form (DAG via ``run_workflow``, nested ensemble, …). Default = one ``run()``.
ReactorTurn = Callable[["ReactorContext"], Awaitable[str]]


@dataclass
class ReactivePeerSpec:
    goal: str | None = None
    role: str | None = None
    channels: list[str] = field(default_factory=list)
    # Turn-body seam (the DAG-in-Peer enabler): override this persona's reaction body. Defaults to the
    # session ``react_with``, then to a single ``run()`` agent turn.
    react: ReactorTurn | None = None


@dataclass
class Reaction:
    persona_id: str
    output: str


# make_runner(persona_id, shared) -> RuntimeRunner; shared = {run_group, signal_source, event_stream}
MakeRunner = Callable[[str, dict[str, Any]], "RuntimeRunner"]


class ReactiveSession:
    def __init__(
        self,
        *,
        run_group: RunGroup,
        turn_policy: TurnPolicy,
        make_runner: MakeRunner,
        event_stream: EventStream | None = None,
        signal_gateway: SignalGateway | None = None,
        goal_for: Callable[[str, BlackboardEvent], str] | None = None,
        react_with: ReactorTurn | None = None,
    ) -> None:
        self._run_group = run_group
        self._turn_policy = turn_policy
        self._make_runner = make_runner
        self._event_stream = event_stream or InMemoryEventStream()
        self._gateway = signal_gateway or SignalGateway()
        self._goal_for = goal_for
        self._react_with = react_with
        self._peer_specs: dict[str, ReactivePeerSpec] = {}
        self._runners: dict[str, "RuntimeRunner"] = {}
        self._policy_state: dict[str, Any] = {}

    def add_peer(self, persona_id: str, *, goal: str | None = None, role: str | None = None,
                 channels: list[str] | None = None, react: ReactorTurn | None = None) -> None:
        self._peer_specs[persona_id] = ReactivePeerSpec(goal, role, channels or [], react)

    def peers(self) -> list[str]:
        return list(self._peer_specs.keys())

    def blackboard(self) -> EventStream:
        return self._event_stream

    async def emit(self, payload: Any, *, source: str | None = None,
                   channel: str | None = None, audience: list[str] | None = None) -> list[Reaction]:
        bb = await self._event_stream.append(payload, source=source, channel=channel, audience=audience)
        # Record lineage for all registered peers (idempotent).
        for pid, spec in self._peer_specs.items():
            await self._run_group.budget_store.join(self._run_group.id, GroupMember(pid, spec.role))

        candidates = [
            PeerView(pid, spec.role, spec.channels)
            for pid, spec in self._peer_specs.items()
            if is_visible_to(bb, EventViewer(pid, spec.channels))
        ]
        chosen = self._turn_policy(bb, candidates, self._policy_state)
        if hasattr(chosen, "__await__"):
            chosen = await chosen
        eligible = {p.persona_id for p in candidates}

        reactions: list[Reaction] = []
        for pid in chosen:
            if pid in eligible:
                reactions.append(Reaction(pid, await self._drive_turn(pid, bb)))
        return reactions

    async def interrupt(self, persona_id: str, *, payload: dict | None = None) -> None:
        self._gateway.ingest(RuntimeSignal(
            kind="interrupt", payload=payload or {}, source="gateway",
            signal_type="alert", urgency="critical", recipient=persona_id,
        ))

    async def broadcast(self, *, payload: dict | None = None) -> None:
        self._gateway.ingest(RuntimeSignal(kind="external", payload=payload or {}, source="gateway"))

    def _get_runner(self, persona_id: str) -> "RuntimeRunner":
        runner = self._runners.get(persona_id)
        if runner is None:
            runner = self._make_runner(persona_id, {
                "run_group": self._run_group,
                "signal_source": self._gateway,
                "event_stream": self._event_stream,
            })
            self._runners[persona_id] = runner
        return runner

    async def _drive_turn(self, persona_id: str, event: BlackboardEvent) -> str:
        runner = self._get_runner(persona_id)
        goal = (
            (self._goal_for(persona_id, event) if self._goal_for else None)
            or self._peer_specs[persona_id].goal
            or "React to the latest events on the shared blackboard."
        )
        # Turn-body seam (DAG-in-Peer): per-peer ``react`` wins, then session ``react_with``, else the
        # default single agent turn. The body's runner inherits the shared RunGroup.
        react = self._peer_specs[persona_id].react or self._react_with
        if react is not None:
            return await react(ReactorContext(persona_id, goal, event, runner))
        return await collect_text(runner.run(session_id=persona_id, goal=goal))

    @staticmethod
    async def resume(
        *, run_group: RunGroup, turn_policy: TurnPolicy, make_runner: MakeRunner,
        event_stream: EventStream | None = None,
        peer_specs: dict[str, ReactivePeerSpec] | None = None,
    ) -> "ReactiveSession":
        session = ReactiveSession(
            run_group=run_group, turn_policy=turn_policy,
            make_runner=make_runner, event_stream=event_stream,
        )
        for member in await run_group.budget_store.members(run_group.id):
            spec = (peer_specs or {}).get(member.session_id) or ReactivePeerSpec(role=member.role)
            session._peer_specs[member.session_id] = spec
        return session


def read_recent_tool(event_stream: EventStream, viewer: EventViewer):
    """A ``read_recent`` tool a persona uses to read the shared blackboard, scoped to what it may see."""

    async def read_recent(since_seq: int = -1, channel: str = "") -> str:
        """Read recent blackboard events visible to you (optionally a single channel you subscribe to)."""
        events = await event_stream.read_since(since_seq, viewer)
        if channel:
            events = [e for e in events if e.channel == channel]
        return json.dumps([
            {"seq": e.seq, "source": e.source, "channel": e.channel, "payload": e.payload}
            for e in events
        ])

    return tool(read_recent)
